test/e2e/framework/expect.go

```diff
@@ -35,17 +35,44 @@ func ExpectError(err error, explain ...interface{}) {
 	gomega.ExpectWithOffset(1, err).To(gomega.HaveOccurred(), explain...)
 }
 
+// ExpectErrorExplained expects an error happens, otherwise an exception raises
+// An additional printf-style explanation must be provided because often errors
+// cannot be understood without it. In cases where the error is known to
+// be descriptive enough in all cases, use ExpectError
```

Comment:
My understanding was that we want to get rid of the other function version eventually. Otherwise we end up (after the renaming) with lots of ExpectErrorUnexplained calls and won't know whether those are used intentionally (error is descriptive) or unintentional (author didn't think about it).

To make this very clear, we might also add a "DEPRECATED: Use ExpectError with empty explain string instead." notice to ExpectErrorUnexplained as part of the rename. I just don't know whether that will cause problems with linters.

@lavalamp : correct?

If yes, then we want the support for empty explainFormatStr here.

Reply:
I think that eventually we should try to get rid of the other function - yes.

That said - I'm not sold on allowing "empty string instead". If we allow it, people will just start using it everywhere, as it's easier to use. I would say: let's require it everywhere. Worst case some duplication of information will appear, but in test failure I prefer duplication of information that no-information.

Reply:
>  I'm not sold on allowing "empty string instead"

That's [fine with me](https://github.com/kubernetes/kubernetes/issues/109600#issuecomment-1105671849). It was meant to get more developers on board with this proposal, but if we can reach consensus that the project always wants an explanation then let's go for that.

The implementation then has to cause a test failure when called with an empty string.

Reply:
> If we allow it, people will just start using it everywhere, as it's easier to use. I would say: let's require it everywhere. 

I think that(called with an empty string) can be checked during the pull requests. 
Do we really need to make it necessary in code?

Reply:
Yes. If neither reviewer nor developer remember or know that an empty string is undesirable, code will get in where the string is empty because "it works".

Reply:
All of the testing issues that I brought up in the leads discussion boil down to this: developers already could write better tests, but they don't. How can we make it easier to write good tests and harder to write bad ones?

Making the explanation required and failing tests falls makes it harder to write bad ones (developers have to add some explanation and hopefully will not write nonsense) and easier (less things a reviewer needs to check).

Reply:
If you call the "Explained" version of the test and don't actually pass an explanation, that's worth failing the test about, IMO.

Obviously we can't force people to give good explanations. But if they don't want to give an explanation, we can make it easier to just call the unexplained version of the function, which makes the problem discoverable / searchable.

Reply:
> Otherwise we end up (after the renaming) with lots of ExpectErrorUnexplained calls and won't know whether those are used intentionally (error is descriptive) or unintentional (author didn't think about it).

We will have a gajillion such call sites and it will be a Sisyphean task to fix them all. The benefit of having a named function is that they at least will be easy to find. No unexplained error is desirable.

> DEPRECATED: Use ExpectError with empty explain string instead

No: this makes the problem hard to search for.

Reply:
> In cases where the error is known to
> // be descriptive enough in all cases, use ExpectError

No, please change this, it's impossible to know that. Maybe: "if you're unsure what to put for the explanation, try to explain the error to a hypothetical person reading failure logs without any other context about the test"

Reply:
> > DEPRECATED: Use ExpectError with empty explain string instead
> 
> No: this makes the problem hard to search for.

That was from before we settled (?) on always requiring an explanation.

Now the notice should be: "DEPRECATED: Use ExpectError with a non-empty explanation."

Does that make sense?

Reply:
I think the documentation should recommend using the explanation-giving version of the function. I'm not sure I actually expect us to literally remove the other version, because I think updating the code base is a large task and will take a long time. So, I'm not sure I'd use the word deprecated.

Perhaps we should just recommend the explanation version for the moment, and in a month or two evaluate:
a) Are people putting in good explanations?
b) Are we actually making progress through the codebase?

If the answer to both is "yes" then I'll be 100% on board with deprecating the "unexplained" way.

Reply:
> b) Are we actually making progress through the codebase?

I don't think it should be criteria. The fact we're not making progress may be cause that we don't have good enough motivation (i.e. that is still easy to use the old version).
I'm with @pohly that we should mark that as deprecated as a stronger inventive for people to not using it.

Reply:
I have extended `hack/verify-golangci-lint.sh` in https://github.com/kubernetes/kubernetes/pull/109728 so that it could be invoked with stricter checking of new code in a PR. I'm mentioning it here first because you guys might be interested. Once I have some initial (hopefully positive :grin:) feedback I can also advertise it more widely (leads or dev mailing list).
 
That currently doesn't have a check for non-empty explanation, but it wouldn't be hard to write a golangci-lint plugin for the specific checks that we care about in Kubernetes.

Reply:
That sounds great - I will try to look at that PR later this week.

I think it would be extremely useful to see how we can extend that to k8s-specific rule (e.g. for this particular thing) as it would be a great motivation for making the case for that and increasing healthiness of our codebase over time.

@dims  FYI ^^

Reply:
I have already integrated an enhanced logcheck linter into our golangci-lint invocation. That works via the plugin mechanism.

Writing more custom linters will definitely be possible. That was my original motivation, and then I realized that just enabling the "normal" golangci-lint would already bring some benefit.

Reply:
> feedback I can also advertise it more widely (leads or dev mailing list).

I'm not in lead mailing list, please use dev so I can follow up

Reply:
all the decisions we can take here can be reverted, so I would be inclined to start with the linter first because IMHO it has a faster feedback loop ... after evaluating it, we have this as secondary option

Reply:
+1